### PR TITLE
[Impactful Change] Multiplayer CompileDef based Client/Server split

### PR DESCRIPTION
**Pull request by AMZN-puvvadar**
## What does this PR do?

This change enables the usage of macros (AZ_TRAIT_CLIENT and AZ_TRAIT_SERVER) to conditionally exclude logic in the Multiplayer Gem such that ServerLaunchers can be built without Client specific code and vice versa. The macro usage approach _also_ allows for a Unified mode and adds a UnifiedLauncher in which all logic is included. This is important for tools and to support Client as a Host based Launchers and allows developers to easily build all target types they may require.

These changes are based on the proposal in this RFC: https://github.com/o3de/sig-network/issues/77

Specifically:
1. Updates LyAutoGen such that multiple targets can specify the same codegen rules. This allows usage of the same ruleset/inputs with different compile flags. This does result in more generated output though this should not impact exe size. This change is **Impactful** as it affects the output path of generated code. It does not however, affect include paths used in code which keeps in code includes the same even with the split.
2. Split Multiplayer into several targets

- Common: All code in Multiplayer that is not split or does not depend on split code
- Client: Common + split code compiled for client usage
- Server: Common + split code compiled for server usage
- Unified: All code, intended for client host usage

3. Expected usage of AZ_TRAIT_CLIENT and AZ_TRAIT_SERVER to test changes and preview how split Multiplayer logic should be authored.
4. Treats Authority to Server NetworkInputs as always non scriptable. I did not want to have platform specific behavior context reflection and it felt unlikely that a user would want to script Auth to Server.
5. Adds a UnifiedLauncher target for users that may want a build featuring Client and Server logic (Client as a Host builds). All Gems have had their cmakes updated appropriately This particular change is **Impactful**.

Note that the intent of ServerLauncher in this prototype is to move it towards being headless.

Feedback on this approach is appreciated. This works but can we do better?
1. Can the cmake approach be simplified at all?
2. Is there a better approach with respect to CodeGen?
3. Is the usability of the Multiplayer API 'good'? Is the tradeoff of security vs usability acceptable as proposed?

## What should reviewers be looking at?
1. CMake, LauncherUnified and project Template changes to add the UnifiedLauncher
2. Changes to the Multiplayer Gem
3. Changes to the Multiplayer Scripting sub Gem. It uses the Multiplayer.Common target as it does not require MP components (which motivates most split logic).
4. Changes to MultiplayerSample: https://github.com/o3de/o3de-multiplayersample/pull/200 which uses all split targets as it uses and builds upon Multiplayer's split components.

All Gems were updated to support the new UnifiedLauncher target. There isn't much to see there.

## How was this PR tested?

There were a few main methods of testing.
1. Functional testing by baselining MPS behavior and ensuring behavior stayed consistent with changes
5. Breakpoint based testing to ensure logic is in fact compiled out appropriately.
6. Inspecting module dependencies to ensure dependency chains made sense.

### CODE REVIEW

**Comment by AMZN-Gene** [comment]
Cool! I think MPS probably has all the various RPC and NetProperty TO->FROM combos, but just in case you might create a temporary XML to test all the jinja combos.

Valid Network property combos
'Authority' to 'Server'
'Authority' to 'Autonomous' 
'Authority' to 'Client'
'Autonomous' to 'Authority' 

Valid RPC combo
'Authority' to 'Autonomous'
'Authority' to 'Client'
'Server' to 'Authority' 
'Autonomous' to 'Authority'

**Review by AMZN-Gene — APPROVED** [review]

File: Gems/Multiplayer/Code/CMakeLists.txt
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/CMakeLists.txt:43]: This multiplayer_files.cmake needs to be as small as possible.
We want to avoid compiling the same file with multiple different compiler flags as much as possible.

Actually per PAL we should use separate files instead of macros

Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/CMakeLists.txt:43]: I can examine the file sets to optimize it further. This isn't a very complete attempt in that regard.

There's definitely reduction that can happen here. The pro of the AzAutoGen changes here is an alternate ruleset specific to each target type can be generated. Though that to me seems functionally equivalent to macro usage with the added maintenance cost of jinja variations. Is there a tradeoff here outside of preventing misuse of the macros? 

Further downstream, I want to avoid exploding out Multiplayer as much as possible. I took a look at separating out into distinct files initially and the Unified requirement makes it challenging.

File: Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Source.jinja
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Source.jinja:1788]: Be careful with these whitespace changes.. you really need to validate each and every case generates the expected formatting. I went through and undid a bunch of these indentation changes recently because the generated files were all wrong.

File: Gems/Multiplayer/Code/CMakeLists.txt
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/CMakeLists.txt:62]: I see you have the autogen target above now, doesn't that mean you can remove the autogen regex expressions from the other targets?

File: Gems/Multiplayer/Code/CMakeLists.txt
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/CMakeLists.txt:12]: Static seems like the wrong thing to use? This should be a headers only or something similar most likely?

**Review by kberg-amzn — COMMENTED** [review]
Lots of good stuff here :thumbsup: nice start!

File: Gems/Multiplayer/Code/CMakeLists.txt
Comment (kberg-amzn) [reply Gems/Multiplayer/Code/CMakeLists.txt:43]: I don't think different files are going to be viable here. We really need the flexibility to go from client to server to unified to hit all our customer use-cases; and we need to do so without convoluting our class architecture and making it harder to author multiplayer content than it already is. Refer to the original discussion we had between you and me and Esteban and a few others.

File: Gems/Multiplayer/Code/CMakeLists.txt
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/CMakeLists.txt:62]: No they're actually doing different thing. AutoGen is providing all the non-split codegen work (Packets mostly). The other targets are doing Components. This is so the set of split logic we're generating per platform is smaller.

File: Gems/Multiplayer/Code/CMakeLists.txt
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/CMakeLists.txt:12]: This is the exact type of thing I'd love the cmake maintainers to weigh in on. It's true this target shouldn't need to be built standalone but it's also not purely headers due to AutoGen so I'm not entirely sure.

Comment (lumberyard-employee-dm) [reply Gems/Multiplayer/Code/CMakeLists.txt:12]: Well an interface targer can hide the a private static target.

Such as having an `Multiplayer.AutoGen.Private.Object STATIC` static target and then having an INTERFACE target that depends on it as `Multiplayer.AutoGen.API INTERFACE`.

Arguably it would be good to split the AutoGen headers and source files up into separate directories or if that is too much work, separate `*_files.cmake` files that contain only public AutoGen API headers` and, so that any the `Multiplayer.AutoGen.API` target can be header only and doesn't have to depend on the static library.

File: Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkRigidBodyComponent.h
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkRigidBodyComponent.h:66]: Something such as this `#if` block technically doesn't need to be in the header.
The #if statement only needs to be in the cpp file.

This function can even be deleted within the Client .cpp file to make it explicit to the user that this function isn't implemented.

```
#if AZ_TRAIT_SERVER_ENABLED
    void NetworkRigidBodyComponentController::HandleSendApplyImpulse(AzNetworking::IConnection* invokingConnection, const AZ::Vector3& impulse, const AZ::Vector3& worldPoint)
    {
       // server implementation
    }
#else
    // Only the server can send impulse to a rigid body
    void NetworkRigidBodyComponentController::HandleSendApplyImpulse(AzNetworking::IConnection* invokingConnection, const AZ::Vector3& impulse, const AZ::Vector3& worldPoint) = delete;
#endif
```

Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkRigidBodyComponent.h:66]: I'll look at querying the compile defs from jinja to do this next. It would definitely be a lot clearer.

**Comment by AMZN-puvvadar** [comment]
I've updated the AutoGen templates and downstream logic for split NetworkInputs and NetworkProperties. Notes are in the description.

File: Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkRigidBodyComponent.h
Comment (kberg-amzn) [reply Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkRigidBodyComponent.h:66]: No, I wouldn't do that, if I understand what you're thinking about correctly..

I would keep the separation between jinja and C++ compilation total and complete, as it is now. Trying to route preprocessor definitions into jinja just for this one use-case (which can totally be implemented without that coupling) seems really confusing and error prone. For an end customer, what's the difference between a preprocessor macro passed through cmake vs one defined in a header file somewhere. We'd have to clearly explain concepts like this to our customers so they understand why some preprocessor values work in their jinja code and others seem to be lost or forgotten. What about changing between debug and profile and release builds? this would change the preprocessor macros, so we'd have to regenerate all autogen outputs which would cause all autogen files to rebuild whenever you switch settings like this. What about different projects that use different preprocessor macros, but link a static lib containing autogen code compiled in a different project? Those are just a few edge cases I can think of off the top of my head. It's surprising behaviour, and it could complicate the build pipeline a lot.

File: Gems/Multiplayer/Code/CMakeLists.txt
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/CMakeLists.txt:119]: This is only added a dependency of the `Multiplayer` target and that target is not used by anything.

So both this target and `Multiplayer` targets should removed

**Comment by AMZN-puvvadar** [comment]
Latest change adds a UnifiedLauncher target and minimal changes for it to function. I have changes adding Unified to all existing Gems but have not yet committed it for brevity. The UnifiedLauncher provides both Client and Server logic to support Client as a Host. It functions similarly to what ServerLauncher has been up to this point with the intention of moving ServerLauncher to a headless process long term.

Testing shows it correctly functioning as either a Client or a Server.

File: Gems/Multiplayer/Code/CMakeLists.txt
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/CMakeLists.txt:42]: I've opted to remove the AutoGen target and just have each target own everything. This leaves the door open to split or other macro based logic in those templates and cleans the CMake up a bit..

File: Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:138]: I'm not the biggest fan of this. I think it does a number on the readability but it does separate the logic nicely. An alternative here is defining UpdateBankedTime and UpdateAutonomous as empty on their unimplemented platforms vs undeclared and keeping the events in.

This could potentially be an option at the CodeGen level.

**Comment by AMZN-Gene** [comment]
Building the INSTALL target on this branch, but there were a few compiler errors in NetworkTransformComponent
`cmake --build ./build/windows_vs2019 --target INSTALL --config profile`

image

**Comment by AMZN-puvvadar** [comment]
> Heads up, building using this branch has a few compiler errors in NetworkTransformComponent. Is this PR maybe missing that file?
> 
> image

It looks like that RPC was added while this review was sitting. What's happening here is the Handler is only defined on one of client or server. You'll want to look in the autogen'd base to see which and then similarly ifdef out these functions.

**Comment by AMZN-Gene** [comment]
> > Heads up, building using this branch has a few compiler errors in NetworkTransformComponent. Is this PR maybe missing that file?
> > image
> 
> It looks like that RPC was added while this review was sitting. What's happening here is RPCs are sent on a specified start endpoint and then handled on a different end endpoint. Autogen will define out relevant functions based on that. i.e. if it's HandleOn Authority, the Handle On will only be defined on Server.
> 
> You'll want to look in the autogen'd base to see which and then similarly ifdef out these functions. There's no function to override here because the base is defined out. I'll need to get this branch merged and updated in the future so I can commit it but it'd great if you can run through it locally for the purposes of feedback.

I have a branch with the fixes, but it also contains dev latest (with Multiplayer Script Gem) so the PR from that branch to this one wont look good; if you want to merge latest dev into this branch, that might help. (or I can)

**Comment by AMZN-puvvadar** [comment]
The PR has been updated against latest dev. Multiplayer_ScriptCanvas presents a case that requires updating. Its cmake should not be as complex as I've had to make it. I'll be looking at adding a Multiplayer.Common.Static it can use.

File: Gems/Multiplayer/Multiplayer_ScriptCanvas/Code/CMakeLists.txt
Comment (AMZN-Gene) [inline Gems/Multiplayer/Multiplayer_ScriptCanvas/Code/CMakeLists.txt:23]: nice!

**Review by AMZN-Gene — APPROVED** [review]

**Review by lsemp3d — APPROVED** [review]

File: Templates/MinimalProject/Template/Gem/CMakeLists.txt
Comment (AMZN-Gene) [inline Templates/MinimalProject/Template/Gem/CMakeLists.txt:85]: +1 for updating templates :D

File: Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Common.jinja
Comment (AMZN-Gene) [inline Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Common.jinja:278]: No need to check "ReplicateFrom". Network Inputs only have 3 variables: name, type and init. NetInput always implicitly replicates from Autonomous->Authority

**Review by AMZN-Gene — APPROVED** [review]

File: Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Common.jinja
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Common.jinja:278]: Yeah, network inputs are always auto to auth.

**Review by AMZN-Olex — APPROVED** [review]

File: AutomatedTesting/Gem/Code/CMakeLists.txt
Comment (AMZN-alexpete) [inline AutomatedTesting/Gem/Code/CMakeLists.txt:60]: nit: Linux likes to have newlines at end of files, worth doing a quick pass to add these where missing.

File: Code/LauncherUnified/Unified.cpp
Comment (AMZN-alexpete) [inline Code/LauncherUnified/Unified.cpp:9]:
```suggestion
```
doesn't look like `string_views` are used in this file

File: Gems/Gestures/Code/CMakeLists.txt
Comment (AMZN-alexpete) [inline Gems/Gestures/Code/CMakeLists.txt:46]: Seems odd to me that Gestures Gem would compile for servers - isn't it for mobile gesture input?

File: Gems/HttpRequestor/Code/CMakeLists.txt
Comment (AMZN-alexpete) [inline Gems/HttpRequestor/Code/CMakeLists.txt:51]: I wonder if it's worth creating a function to forward all known aliases for gems like these where every alias is the same e.g. 
```cmake
ly_create_default_alias(NAMESPACE Gem Targets Gem::HttpRequestor)
```

File: Gems/Multiplayer/Code/CMakeLists.txt
Comment (AMZN-alexpete) [inline Gems/Multiplayer/Code/CMakeLists.txt:179]: do you need this `RUNTIME_DEPENDENCIES` entry if the static library already has it?

File: Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp
Comment (AMZN-alexpete) [inline Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:29]: Thoughts on adding/using AZ_CLIENT_CVAR/AZ_SERVER_CVAR instead of #if AZ_TRAIT_SERVER/#if AZ_CLIENT_SERVER?

File: Gems/StartingPointMovement/Code/CMakeLists.txt
Comment (AMZN-alexpete) [inline Gems/StartingPointMovement/Code/CMakeLists.txt:28]: nit: another missing newline

File: Gems/ScriptedEntityTweener/Code/CMakeLists.txt
Comment (AMZN-alexpete) [inline Gems/ScriptedEntityTweener/Code/CMakeLists.txt:46]: nit: missing newline

File: AutomatedTesting/Gem/Code/CMakeLists.txt
Comment (lumberyard-employee-dm) [reply AutomatedTesting/Gem/Code/CMakeLists.txt:60]: We can add a pre-commit hook or update .editorconfig in the repo to automatically insert the final newline.
https://stackoverflow.com/questions/60661412/git-error-if-no-newline-at-end-of-file

**Review by amzn-changml — APPROVED** [review]
From the build side, this seems reasonable. As there are some changes to codegen and headers, there might be stale references in incremental builds. 

We'll schedule a clean build in the dev branch CI after this gets merged to ensure that stale artifacts are kept to a minimum

File: Gems/Gestures/Code/CMakeLists.txt
Comment (AMZN-puvvadar) [reply Gems/Gestures/Code/CMakeLists.txt:46]: Yes, changes like this will be saved for a future change.

File: Gems/HttpRequestor/Code/CMakeLists.txt
Comment (AMZN-puvvadar) [reply Gems/HttpRequestor/Code/CMakeLists.txt:51]: I had this thought and it's hard to say where that would begin or end. There's quite a few aliases that can be specified. This change may begin a trend that makes a default set even less likely.

File: Gems/Multiplayer/Code/CMakeLists.txt
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/CMakeLists.txt:179]: The static library doesn't have it. Are you looking at the Multiplayer target? That's the Unified version of this target. I kept it Multiplayer instead of Multiplayer.Unified for compatibility purposes.

File: Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:29]: Are you suggesting a runtime cvar? We definitely want this to be a compile time determination.

Comment (kberg-amzn) [reply Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:29]: Yeah, removing server only code from client binaries is a critical must-need for MMOs, free to play titles, and anything else where preventing 3rd parties standing up grey servers is a business critical requirement.

**Review by SergeyAMZN — APPROVED** [review]

File: Gems/Multiplayer/Code/CMakeLists.txt
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/CMakeLists.txt]: Yes, they can. Good eye!

File: Gems/Multiplayer/Code/Include/Multiplayer/Components/LocalPredictionPlayerInputComponent.h
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Include/Multiplayer/Components/LocalPredictionPlayerInputComponent.h:102]: Not terribly important, but maybe coalesce the client and server member sections just to minimize the amount of #if / #endif 's we have in the header..

**Review by kberg-amzn — APPROVED** [review]
What a massive amount of work.. nicely done 👍

**Comment by BytesOfPiDev** [comment]
Excited for this one 🥳

**Pull request merged**
